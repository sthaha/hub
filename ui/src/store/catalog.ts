import { Instance, types } from 'mobx-state-tree';

export const Catalog = types
  .model({
    id: types.identifierNumber,
    name: types.optional(types.string, ''),
    type: types.optional(types.string, ''),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICatalog = Instance<typeof Catalog>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model({
    catalogs: types.map(Catalog)
  })

  .actions((self) => ({
    add(item: ICatalog) {
      self.catalogs.put({ id: item.id, name: item.name, type: item.type });
    },
    clearSelected() {
      self.catalogs.forEach((c) => {
        c.selected = false;
      });
    }
  }))

  .views((self) => ({
    get items() {
      return Array.from(self.catalogs.values());
    },

    get selected() {
      const list = new Set();
      self.catalogs.forEach((c: ICatalog) => {
        if (c.selected) {
          list.add(c.id);
        }
      });

      return list;
    }
  }));
