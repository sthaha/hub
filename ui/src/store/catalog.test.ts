import { Catalog, CatalogStore } from './catalog';
import { getSnapshot } from 'mobx-state-tree';

describe('Store Object', () => {
  it('can create a catalog object', () => {
    const store = Catalog.create({
      id: 1,
      name: 'tekton',
      type: 'community'
    });

    expect(store.name).toBe('tekton');
  });

  it('creates a catalog store', (done) => {
    const store = CatalogStore.create({});

    const item = Catalog.create({
      id: 1,
      name: 'tekton',
      type: 'community'
    });

    store.add(item);

    expect(getSnapshot(store.catalogs)).toMatchSnapshot();

    done();
  });

  it('should toggle a selected catalog', (done) => {
    const store = CatalogStore.create({});

    const item = Catalog.create({
      id: 1,
      name: 'tekton',
      type: 'community'
    });

    store.add(item);
    store.catalogs.get('1')?.toggle();

    expect(store.selected.size).toBe(1);
    expect(store.catalogs.get('1')?.selected).toBe(true);

    done();
  });

  it('should clear all the selected catalog', (done) => {
    const store = CatalogStore.create({});

    const item = Catalog.create({
      id: 1,
      name: 'tekton',
      type: 'community'
    });

    store.add(item);
    store.catalogs.get('1')?.toggle();

    store.clearSelected();

    expect(store.catalogs.get('1')?.selected).toBe(false);

    done();
  });
});
