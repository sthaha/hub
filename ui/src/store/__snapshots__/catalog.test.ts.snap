// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Object creates a catalog store 1`] = `
Object {
  "1": Object {
    "id": 1,
    "name": "tekton",
    "selected": false,
    "type": "community",
  },
}
`;
